File: modules/scv-engine/tests/interface_generation_test.py
#!/usr/bin/env python3
"""Independent historical/interface-authoring regression checks."""
import copy
import importlib.util
import json
from pathlib import Path
import subprocess
import sys
import tempfile
import unittest

sys.dont_write_bytecode = True
ROOT = Path(__file__).resolve().parents[3]
SPEC = importlib.util.spec_from_file_location("scv_interface_generator", ROOT / "modules/scv-engine/tools/generate_interface.py")
gen = importlib.util.module_from_spec(SPEC)
SPEC.loader.exec_module(gen)


class InterfaceGeneration(unittest.TestCase):
    def setUp(self):
        self.manifest = gen.read_json(ROOT / gen.MANIFEST)
        self.temp = tempfile.TemporaryDirectory()
        self.addCleanup(self.temp.cleanup)
        self.root = Path(self.temp.name)
        self.write(gen.MANIFEST, self.manifest)
        self.write(gen.HISTORY, gen.read_json(ROOT / gen.HISTORY))
        self.catalog = {"protocol": "symphony.scv.schema-catalog.v1", "engine_version": self.manifest["current_release"], "entries": []}
        definitions = {}
        for op in self.manifest["operations"]:
            for field, kind in (("input_protocol", "input"), ("output_protocol", "output")):
                protocol = op[field]
                definitions[protocol] = {"type": "object"}
                self.catalog["entries"].append({"protocol": protocol, "kind": kind, "operations": [op["name"]], "file": "fixture.schema.json", "fragment": "#/$defs/" + protocol})
        self.write(self.manifest["schema_catalog"], self.catalog)
        self.write("knowledge/scv/schemas/v1/fixture.schema.json", {"$defs": definitions})

    def write(self, name, value):
        path = self.root / name
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(json.dumps(value))

    def test_historical_projection_is_frozen(self):
        gen.check_history(self.manifest, ROOT)
        self.assertEqual([len(gen.projection(self.manifest, f"0.{n}.0-dev")["operations"]) for n in range(1, 7)], [13, 17, 20, 21, 22, 26])
        self.manifest["operations"][1]["output_protocol"] = "symphony.scv.changed.v1"
        with self.assertRaisesRegex(gen.Invalid, "frozen interface changed"):
            gen.check_history(self.manifest, ROOT)

    def test_future_domain_does_not_change_old_admission_or_definition(self):
        original = copy.deepcopy(self.manifest)
        next_release = copy.deepcopy(self.manifest["releases"][-1])
        next_release["version"] = "0.7.0-dev"
        self.manifest["releases"].append(next_release)
        self.manifest["current_release"] = "0.7.0-dev"
        self.manifest["domains"].append({"name": "scev-new-provider", "introduced_in": "0.7.0-dev"})
        gen.validate(self.manifest)
        for release in original["releases"]:
            version = release["version"]
            self.assertEqual(gen.projection(original, version), gen.projection(self.manifest, version))
            self.assertEqual(gen.definition_digest(gen.release_manifest(original, version)), gen.definition_digest(gen.release_manifest(self.manifest, version)))
        self.assertIn("scev-new-provider", gen.projection(self.manifest, "0.7.0-dev")["domains"])

    def test_complete_generation_is_deterministic(self):
        first = gen.render(self.manifest, self.root)
        self.assertEqual(first, gen.render(self.manifest, self.root))
        self.assertEqual(set(first), {gen.CPP_OUTPUT, gen.GO_OUTPUT, gen.CMAKE_OUTPUT})
        self.assertIn(b"OWNER-INTERFACE.json", first[gen.CMAKE_OUTPUT])
        self.assertIn(b"fixture.schema.json", first[gen.CMAKE_OUTPUT])

    def test_check_detects_output_drift(self):
        command = [sys.executable, str(ROOT / "modules/scv-engine/tools/generate_interface.py"), "--root", str(self.root)]
        self.assertEqual(subprocess.run(command, capture_output=True).returncode, 0)
        self.assertEqual(subprocess.run(command + ["--check"], capture_output=True).returncode, 0)
        path = self.root / gen.GO_OUTPUT
        path.write_bytes(path.read_bytes() + b"// drift\n")
        result = subprocess.run(command + ["--check"], capture_output=True)
        self.assertNotEqual(result.returncode, 0)
        self.assertIn(b"generated interface drift", result.stderr)

    def test_malformed_manifest_rejected(self):
        mutations = [
            lambda m: m.update(unknown=True),
            lambda m: m.update(current_release="0.7.0-dev"),
            lambda m: m["domains"].append(copy.deepcopy(m["domains"][0])),
            lambda m: m["domains"][0].update(introduced_in="unknown"),
            lambda m: m["operations"].append(copy.deepcopy(m["operations"][0])),
            lambda m: m["operations"][0].update(handler="run_shell"),
            lambda m: m["operations"][0].update(mutability="unrestricted"),
            lambda m: m["operations"][0].update(expected_state=1),
            lambda m: m["operations"][0].update(interactions=[{}]),
            lambda m: m["operations"][-1]["artifact"].update(introduced_in="0.1.0-dev"),
            lambda m: m["releases"][0].update(companions=[{}]),
        ]
        for mutation in mutations:
            with self.subTest(mutation=mutations.index(mutation)):
                value = copy.deepcopy(self.manifest)
                mutation(value)
                with self.assertRaises(gen.Invalid):
                    gen.validate(value)

    def test_malformed_catalog_and_references_rejected(self):
        mutations = [
            lambda c: c.update(engine_version="0.5.0-dev"),
            lambda c: c["entries"].pop(0),
            lambda c: c["entries"].append(copy.deepcopy(c["entries"][0])),
            lambda c: c["entries"].append(4),
            lambda c: c["entries"][0].update(kind="input" if c["entries"][0]["kind"] == "output" else "output"),
            lambda c: c["entries"][0].update(file="../escape.schema.json"),
            lambda c: c["entries"][0].update(fragment="#/$defs/missing"),
        ]
        for i, mutation in enumerate(mutations):
            with self.subTest(mutation=i):
                value = copy.deepcopy(self.catalog)
                mutation(value)
                self.write(self.manifest["schema_catalog"], value)
                with self.assertRaises(gen.Invalid):
                    gen.schema_inventory(self.manifest, self.root)
        self.write(self.manifest["schema_catalog"], self.catalog)
        path = "knowledge/scv/schemas/v1/fixture.schema.json"
        doc = gen.read_json(self.root / path)
        for ref in ("https://example.invalid/schema", "../escape.schema.json", "#/$defs/missing", "#/$defs/~2invalid"):
            with self.subTest(reference=ref):
                value = copy.deepcopy(doc)
                value["$ref"] = ref
                self.write(path, value)
                with self.assertRaises(gen.Invalid):
                    gen.schema_inventory(self.manifest, self.root)

    def test_duplicate_json_keys_rejected(self):
        path = self.root / "duplicate.json"
        path.write_text('{"key":1,"key":2}')
        with self.assertRaises(gen.Invalid):
            gen.read_json(path)

    def test_checked_in_metadata_has_no_drift(self):
        for name, expected in gen.render(self.manifest, ROOT, metadata_only=True).items():
            self.assertEqual((ROOT / name).read_bytes(), expected, name)

    def test_artifact_schema_admissions_match_owner_interface(self):
        definitions = gen.read_json(ROOT / "knowledge/scv/schemas/v1/scv-artifact.schema.json")["$defs"]
        artifacts = [op for op in self.manifest["operations"] if op["artifact"]]
        operations = [op["name"] for op in artifacts]
        kinds = [op["artifact"]["kind"] for op in artifacts]
        self.assertEqual(definitions["ImportInput"]["properties"]["operation"]["enum"], operations)
        self.assertEqual(definitions["Record"]["properties"]["operation"]["enum"], operations)
        self.assertEqual(definitions["Record"]["properties"]["kind"]["enum"], kinds)
        catalog = gen.read_json(ROOT / self.manifest["schema_catalog"])
        entries = {entry["protocol"]: entry for entry in catalog["entries"]}
        versions = [release["version"] for release in self.manifest["releases"]]
        def reference(protocol):
            entry = entries[protocol]
            return entry["file"] + entry["fragment"]
        self.assertEqual(len(definitions["Record"]["allOf"]), len(artifacts))
        self.assertEqual(len(definitions["ImportInput"]["allOf"]), len(artifacts))
        for op, imported, recorded in zip(artifacts, definitions["ImportInput"]["allOf"], definitions["Record"]["allOf"]):
            for branch in (imported, recorded):
                self.assertEqual(branch["if"]["properties"]["operation"]["const"], op["name"])
                self.assertEqual(branch["then"]["properties"]["input"], {"$ref": reference(op["input_protocol"])})
            self.assertEqual(imported["then"]["properties"]["result"], {"anyOf": [{"type": "null"}, {"$ref": reference(op["output_protocol"])}]})
            properties = recorded["then"]["properties"]
            self.assertEqual(properties["artifact"], {"$ref": reference(op["output_protocol"])})
            self.assertEqual(properties["kind"]["const"], op["artifact"]["kind"])
            admitted = [version for version in versions if gen.projection(self.manifest, version)["operations"] and any(x["name"] == op["name"] and x["artifact_kind"] for x in gen.projection(self.manifest, version)["operations"])]
            self.assertEqual(properties["installation"]["properties"]["Version"]["enum"], admitted)

    def test_checked_in_complete_inventory_has_no_drift(self):
        # This is deliberately a full check, not the incomplete bootstrap mode.
        for name, expected in gen.render(self.manifest, ROOT).items():
            self.assertEqual((ROOT / name).read_bytes(), expected, name)


if __name__ == "__main__":
    unittest.main()

File: modules/scv-engine/tests/interface_generation_test.cpp
#include "scv_interface.hpp"
#include "test_support.hpp"
using namespace symphony::authoring;
struct InterfaceGeneration {
  Arguments args;
  Json manifest;
  TempDir temp;
  fs::path root;
  Json catalog;
  std::string catalog_path;
  const std::string fixture_path =
      "knowledge/scv/schemas/v1/fixture.schema.json";
  explicit InterfaceGeneration(const Arguments &selected)
      : args(selected), manifest(read_json(args.root / scv::manifest_path)),
        root(temp.path), catalog_path(string(manifest.at("schema_catalog"))) {
    write_json(root / scv::manifest_path, manifest);
    test::copy(args.root, root, scv::history_path);
    catalog = {{"protocol", "symphony.scv.schema-catalog.v1"},
               {"engine_version", manifest.at("current_release")},
               {"entries", Json::array()}};
    Json definitions = Json::object();
    for (const auto &op : manifest.at("operations"))
      for (const auto &[field, kind] :
           std::vector<std::pair<std::string, std::string>>{
               {"input_protocol", "input"}, {"output_protocol", "output"}}) {
        const auto protocol = string(op.at(field));
        definitions[protocol] = {{"type", "object"}};
        catalog["entries"].push_back(
            {{"protocol", protocol},
             {"kind", kind},
             {"operations", Json::array({op.at("name")})},
             {"file", "fixture.schema.json"},
             {"fragment", "#/$defs/" + protocol}});
      }
    write_json(root / catalog_path, catalog);
    write_json(root / fixture_path, {{"$defs", definitions}});
  }
  void test_historical_projection_is_frozen() {
    scv::check_history(manifest, args.root);
    const std::vector<int> counts{13, 17, 20, 21, 22, 26, 26, 28, 30, 30};
    for (int n = 1; n <= 10; ++n)
      test::check(scv::projection(manifest, "0." + std::to_string(n) + ".0-dev")
                          .at("operations")
                          .size() == static_cast<std::size_t>(counts.at(n - 1)),
                  "historical operation count");
    auto changed = manifest;
    changed["operations"][1]["output_protocol"] = "symphony.scv.changed.v1";
    test::rejects([&] { scv::check_history(changed, args.root); },
                  "frozen projection", "frozen interface changed");
  }
  void test_malformed_history_release_inventory_rejected() {
    const auto history = read_json(root / scv::history_path);
    for (const auto &invalid :
         Json::array({nullptr,
                      Json{{"arbitrary-key", history.at("releases").at(0)}}})) {
      auto changed = history;
      changed["releases"] = invalid;
      write_json(root / scv::history_path, changed);
      test::rejects([&] { scv::check_history(manifest, root); },
                    "malformed frozen release inventory",
                    "frozen history releases must be an array");
    }
  }
  void test_future_domain_does_not_change_old_admission_or_definition() {
    auto future = manifest;
    auto release = future.at("releases").back();
    release["version"] = "0.11.0-dev";
    future["releases"].push_back(release);
    future["current_release"] = "0.11.0-dev";
    future["domains"].push_back(
        {{"name", "scev-new-provider"}, {"introduced_in", "0.11.0-dev"}});
    scv::validate(future);
    for (const auto &r : manifest.at("releases")) {
      const auto version = r.at("version");
      test::equal(scv::projection(manifest, version),
                  scv::projection(future, version),
                  "future domain preserves previous projection");
      test::check(digest(scv::release_manifest(manifest, version), false) ==
                      digest(scv::release_manifest(future, version), false),
                  "future domain preserves definition digest");
    }
    test::check(contains(scv::projection(future, "0.11.0-dev").at("domains"),
                         "scev-new-provider"),
                "future domain admitted only in new release");
  }
  void frozen_release(int n) {
    const auto version = "0." + std::to_string(n) + ".0-dev";
    const auto old = read_json(
        args.root / ("modules/scv-engine/tests/fixtures/owner-interface-0." +
                     std::to_string(n) + ".v1.json"));
    test::equal(old.at("current_release"), version,
                "frozen declaration version");
    test::equal(scv::release_manifest(manifest, version), old,
                "complete frozen release declaration");
    test::equal(scv::projection(manifest, version),
                scv::projection(old, version),
                "complete frozen release projection");
  }
  void test_sixth_release_declaration_remains_exact() {
    frozen_release(6);
    test::check(!contains(scv::projection(manifest, "0.6.0-dev").at("surfaces"),
                          "composition_workflow"),
                "sixth release excludes later workflow");
    test::check(contains(scv::projection(manifest, "0.7.0-dev").at("surfaces"),
                         "composition_workflow"),
                "seventh release admits workflow");
  }
  void test_seventh_release_declaration_remains_exact() { frozen_release(7); }
  void test_eighth_release_declaration_remains_exact() { frozen_release(8); }
  void test_ninth_release_declaration_remains_exact() {
    frozen_release(9);
    test::check(!contains(scv::projection(manifest, "0.9.0-dev").at("surfaces"),
                          "composition_bundle_workflow"),
                "ninth release excludes later bundle workflow");
    test::check(contains(scv::projection(manifest, "0.10.0-dev").at("surfaces"),
                         "composition_bundle_workflow"),
                "tenth release admits bundle workflow");
  }
  void test_complete_generation_is_deterministic() {
    const auto output = scv::render(manifest, root);
    test::check(output == scv::render(manifest, root),
                "deterministic complete generation");
    test::check(output.size() == 3 && output[0].first == scv::cpp_path &&
                    output[1].first == scv::go_path &&
                    output[2].first == scv::cmake_path,
                "complete output inventory");
    test::check(
        output[2].second.find("OWNER-INTERFACE.json") != std::string::npos &&
            output[2].second.find("fixture.schema.json") != std::string::npos,
        "CMake inventory names owner and schema");
  }
  void test_check_detects_output_drift() {
    const std::vector<std::string> command{SCV_GENERATOR_EXECUTABLE, "--root",
                                           root.string()};
    const auto generated = run(command);
    test::check(generated.status == 0,
                "generator CLI succeeds: " + generated.error);
    auto check = command;
    check.push_back("--check");
    test::check(run(check).status == 0, "generator CLI exact check succeeds");
    write_bytes(root / scv::go_path,
                read_bytes(root / scv::go_path) + "// drift\n");
    const auto result = run(check);
    test::check(result.status != 0 &&
                    result.error.find("generated interface drift") !=
                        std::string::npos,
                "generator CLI drift rejection");
  }
  void test_malformed_manifest_rejected() {
    const std::vector<std::function<void(Json &)>> mutations{
        [](Json &m) { m["unknown"] = true; },
        [](Json &m) { m["current_release"] = "0.11.0-dev"; },
        [](Json &m) {
          const auto d = m["domains"][0];
          m["domains"].push_back(d);
        },
        [](Json &m) { m["domains"][0]["introduced_in"] = "unknown"; },
        [](Json &m) {
          const auto op = m["operations"][0];
          m["operations"].push_back(op);
        },
        [](Json &m) { m["operations"][0]["handler"] = "run_shell"; },
        [](Json &m) { m["operations"][0]["mutability"] = "unrestricted"; },
        [](Json &m) { m["operations"][0]["expected_state"] = 1; },
        [](Json &m) {
          m["operations"][0]["interactions"] = Json::array({Json::object()});
        },
        [](Json &m) {
          m["operations"].back()["artifact"]["introduced_in"] = "0.1.0-dev";
        },
        [](Json &m) {
          m["releases"][0]["companions"] = Json::array({Json::object()});
        }};
    for (std::size_t i = 0; i < mutations.size(); ++i) {
      auto bad = manifest;
      mutations[i](bad);
      test::rejects([&] { scv::validate(bad); },
                    "malformed declaration " + std::to_string(i));
    }
  }
  void test_malformed_catalog_and_references_rejected() {
    const std::vector<std::function<void(Json &)>> mutations{
        [](Json &c) { c["engine_version"] = "0.5.0-dev"; },
        [](Json &c) { c["entries"].erase(0); },
        [](Json &c) {
          const auto entry = c["entries"][0];
          c["entries"].push_back(entry);
        },
        [](Json &c) { c["entries"].push_back(4); },
        [](Json &c) {
          c["entries"][0]["kind"] =
              c["entries"][0]["kind"] == "output" ? "input" : "output";
        },
        [](Json &c) { c["entries"][0]["file"] = "../escape.schema.json"; },
        [](Json &c) { c["entries"][0]["fragment"] = "#/$defs/missing"; }};
    for (std::size_t i = 0; i < mutations.size(); ++i) {
      auto bad = catalog;
      mutations[i](bad);
      write_json(root / catalog_path, bad);
      test::rejects([&] { scv::schema_inventory(manifest, root); },
                    "malformed catalog " + std::to_string(i));
    }
    write_json(root / catalog_path, catalog);
    const auto doc = read_json(root / fixture_path);
    for (const auto *ref :
         {"https://example.invalid/schema", "../escape.schema.json",
          "#/$defs/missing", "#/$defs/~2invalid"}) {
      auto bad = doc;
      bad["$ref"] = ref;
      write_json(root / fixture_path, bad);
      test::rejects([&] { scv::schema_inventory(manifest, root); },
                    "schema reference rejection");
    }
  }
  void test_duplicate_json_keys_rejected() {
    write_bytes(root / "duplicate.json", "{\"key\":1,\"key\":2}");
    test::rejects([&] { read_json(root / "duplicate.json"); },
                  "duplicate JSON keys");
  }
  void test_checked_in_metadata_has_no_drift() {
    emit(args.root, scv::render(manifest, args.root, true), true);
  }
  void test_checked_in_complete_inventory_has_no_drift() {
    emit(args.root, scv::render(manifest, args.root), true);
  }
  void test_artifact_schema_admissions_match_owner_interface() {
    const auto defs =
        read_json(args.root /
                  "knowledge/scv/schemas/v1/scv-artifact.schema.json")
            .at("$defs");
    Json artifacts = Json::array(), operation_names = Json::array(),
         kinds = Json::array();
    for (const auto &op : manifest.at("operations"))
      if (!op.at("artifact").is_null()) {
        artifacts.push_back(op);
        operation_names.push_back(op.at("name"));
        kinds.push_back(op.at("artifact").at("kind"));
      }
    test::equal(
        defs.at("ImportInput").at("properties").at("operation").at("enum"),
        operation_names, "artifact import operations");
    test::equal(defs.at("Record").at("properties").at("operation").at("enum"),
                operation_names, "artifact record operations");
    test::equal(defs.at("Record").at("properties").at("kind").at("enum"), kinds,
                "artifact record kinds");
    const auto real_catalog = read_json(args.root / catalog_path);
    std::map<std::string, Json> entries;
    for (const auto &entry : real_catalog.at("entries"))
      entries[string(entry.at("protocol"))] = entry;
    const auto reference = [&](const Json &protocol) {
      const auto &entry = entries.at(string(protocol));
      return string(entry.at("file")) + string(entry.at("fragment"));
    };
    test::check(defs.at("Record").at("allOf").size() == artifacts.size() &&
                    defs.at("ImportInput").at("allOf").size() ==
                        artifacts.size(),
                "complete artifact admission branches");
    for (std::size_t i = 0; i < artifacts.size(); ++i) {
      const auto &op = artifacts.at(i);
      const auto &imported = defs.at("ImportInput").at("allOf").at(i);
      const auto &recorded = defs.at("Record").at("allOf").at(i);
      for (const auto &branch : {imported, recorded}) {
        test::equal(
            branch.at("if").at("properties").at("operation").at("const"),
            op.at("name"), "artifact branch identity");
        test::equal(branch.at("then").at("properties").at("input"),
                    {{"$ref", reference(op.at("input_protocol"))}},
                    "artifact input schema");
      }
      test::equal(
          imported.at("then").at("properties").at("result"),
          {{"anyOf",
            Json::array({{{"type", "null"}},
                         {{"$ref", reference(op.at("output_protocol"))}}})}},
          "artifact imported result schema");
      const auto &properties = recorded.at("then").at("properties");
      test::equal(properties.at("artifact"),
                  {{"$ref", reference(op.at("output_protocol"))}},
                  "artifact output schema");
      test::equal(properties.at("kind").at("const"),
                  op.at("artifact").at("kind"), "artifact kind identity");
      Json admitted = Json::array();
      for (const auto &version : scv::versions(manifest)) {
        const auto projected = scv::projection(manifest, version);
        for (const auto &p : projected.at("operations"))
          if (p.at("name") == op.at("name") && !p.at("artifact_kind").is_null())
            admitted.push_back(version);
      }
      test::equal(properties.at("installation")
                      .at("properties")
                      .at("Version")
                      .at("enum"),
                  admitted, "exact artifact release admissions");
    }
  }
};
int main(int argc, char **argv) {
  return main_guard([&] {
    const Arguments args(argc, argv);
    const std::vector<std::pair<std::string, void (InterfaceGeneration::*)()>>
        cases{
            {"test_historical_projection_is_frozen",
             &InterfaceGeneration::test_historical_projection_is_frozen},
            {"test_malformed_history_release_inventory_rejected",
             &InterfaceGeneration::
                 test_malformed_history_release_inventory_rejected},
            {"test_future_domain_does_not_change_old_admission_or_definition",
             &InterfaceGeneration::
                 test_future_domain_does_not_change_old_admission_or_definition},
            {"test_sixth_release_declaration_remains_exact",
             &InterfaceGeneration::
                 test_sixth_release_declaration_remains_exact},
            {"test_seventh_release_declaration_remains_exact",
             &InterfaceGeneration::
                 test_seventh_release_declaration_remains_exact},
            {"test_eighth_release_declaration_remains_exact",
             &InterfaceGeneration::
                 test_eighth_release_declaration_remains_exact},
            {"test_ninth_release_declaration_remains_exact",
             &InterfaceGeneration::
                 test_ninth_release_declaration_remains_exact},
            {"test_complete_generation_is_deterministic",
             &InterfaceGeneration::test_complete_generation_is_deterministic},
            {"test_check_detects_output_drift",
             &InterfaceGeneration::test_check_detects_output_drift},
            {"test_malformed_manifest_rejected",
             &InterfaceGeneration::test_malformed_manifest_rejected},
            {"test_malformed_catalog_and_references_rejected",
             &InterfaceGeneration::
                 test_malformed_catalog_and_references_rejected},
            {"test_duplicate_json_keys_rejected",
             &InterfaceGeneration::test_duplicate_json_keys_rejected},
            {"test_checked_in_metadata_has_no_drift",
             &InterfaceGeneration::test_checked_in_metadata_has_no_drift},
            {"test_checked_in_complete_inventory_has_no_drift",
             &InterfaceGeneration::
                 test_checked_in_complete_inventory_has_no_drift},
            {"test_artifact_schema_admissions_match_owner_interface",
             &InterfaceGeneration::
                 test_artifact_schema_admissions_match_owner_interface}};
    for (const auto &[name, function] : cases) {
      InterfaceGeneration suite(args);
      std::cout << "RUN " << name << '\n';
      (suite.*function)();
      std::cout << "PASS " << name << '\n';
    }
    test::done();
  });
}
